storagenode/gracefulexit/chore.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/storagenode/satellites"
+)
+
+var mon = monkit.Package()
+
+// Chore checks for satellites that are being exited and creates a worker per satellite to complete the process.
+//
+// architecture: Chore
+type Chore struct {
+	log         *zap.Logger
+	satelliteDB satellites.DB
+
+	config Config
+
+	exitingMap sync.Map
+	Loop       sync2.Cycle
+	limiter    sync2.Limiter
+}
+
+// Config for the chore
+type Config struct {
+	ChoreInterval time.Duration `help:"how often to run the chore to check for satellites that need to exit." releaseDefault:"15m" devDefault:"10s"`
+	NumWorkers    int           `help:"number of workers to handle satellite exits" default:"3"`
+}
+
+// NewChore instantiates Chore.
+func NewChore(log *zap.Logger, config Config, satelliteDB satellites.DB) *Chore {
+	return &Chore{
+		log:         log,
+		satelliteDB: satelliteDB,
+		config:      config,
+		Loop:        *sync2.NewCycle(config.ChoreInterval),
+		limiter:     *sync2.NewLimiter(config.NumWorkers),
+	}
+}
+
+// Run starts the chore.
+func (chore *Chore) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	err = chore.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		chore.log.Info("running graceful exit chore.")
+
+		satellites, err := chore.satelliteDB.ListGracefulExits(ctx)
+		if err != nil {
+			chore.log.Error("error retrieving satellites.", zap.Error(err))
+			return nil
+		}
+
+		if len(satellites) == 0 {
+			chore.log.Debug("no satellites found.")
+			return nil
+		}
+
+		for _, satellite := range satellites {
+			satelliteID := satellite.SatelliteID
+			// returns ok == true if the ID existed, otherwise it stores it.
```

Comment:
nit - I feel like this comment can maybe be removed, I personally felt like the `if _, ok` check is clear enough.

Reply:
removed